src/webresource-handler/handlers/S3Handler.ts

```diff
@@ -29,12 +30,55 @@ export interface S3HandlerProps {
 	maxSize?: number;
 	signedUrlExpireTimeSeconds?: number;
 	signedUrlCacheExpireTimeSeconds?: number;
+	calculateChecksums?: boolean;
+}
+
+class ChecksumCalculator {
+	private readonly checksumPromise: Promise<string | undefined>;
+	private readonly passthrough: PassThrough;
+
+	constructor(
+		private readonly calculateChecksums: boolean,
+		private readonly stream: Readable,
+	) {
+		const hash = createHash('sha256');
+
+		this.passthrough = new PassThrough();
+
+		this.checksumPromise = new Promise((resolve) => {
+			if (this.calculateChecksums) {
+				this.stream.on('data', (chunk: Buffer) => {
+					hash.update(chunk);
+					this.passthrough.write(chunk);
+				});
+
+				this.stream.on('end', () => {
+					resolve(hash.digest('hex'));
+					this.passthrough.end();
+				});
```

Comment:
Can this fail? If so, would we like to handle such failure at this point in code?

I am mostly referring to `error` events -- https://nodejs.org/dist/latest-v18.x/docs/api/stream.html

Reply:
Very nice catch @pranasziaukas 
In order to catch all the events and properly pipe them, I updated the code to make use of a pipe, therefore properly piping all the events without having to manually enumerate them.

Now basically we implement the `Transform` interface from node stream api where we basically pipe the data with a side effect which is calculating the checksum. .update() and .digest() methods should not throw, so we are fine just calling them without try/catches. Let me know what do you think!